src/invocation/ClusterService.ts

```diff
@@ -14,317 +14,283 @@
  * limitations under the License.
  */
 
-import {ClientConnection} from './ClientConnection';
+import {ClientConnection} from '../network/ClientConnection';
 import * as Promise from 'bluebird';
-import {ClientAddMembershipListenerCodec} from '../codec/ClientAddMembershipListenerCodec';
 import {Member} from '../core/Member';
-import {LoggingService} from '../logging/LoggingService';
 import {ClientInfo} from '../ClientInfo';
 import HazelcastClient from '../HazelcastClient';
-import {IllegalStateError} from '../HazelcastError';
-import * as assert from 'assert';
+import {IllegalStateError, TargetDisconnectedError} from '../HazelcastError';
 import {MemberSelector} from '../core/MemberSelector';
-import {AddressHelper, DeferredPromise} from '../Util';
-import {MemberAttributeEvent, MemberAttributeOperationType} from '../core/MemberAttributeEvent';
+import {assertNotNull, DeferredPromise} from '../Util';
 import {MembershipListener} from '../core/MembershipListener';
 import {MembershipEvent} from '../core/MembershipEvent';
 import {UuidUtil} from '../util/UuidUtil';
 import {ILogger} from '../logging/ILogger';
-import Address = require('../Address');
-import ClientMessage = require('../ClientMessage');
+import {UUID} from '../core/UUID';
+import {ClientConnectionManager} from '../network/ClientConnectionManager';
+import {InitialMembershipListener} from '../core/InitialMembershipListener';
+import {InitialMembershipEvent} from '../core/InitialMembershipEvent';
+import {MemberInfo} from '../core/MemberInfo';
+import {Cluster} from '../core/Cluster';
 
 export enum MemberEvent {
     ADDED = 1,
     REMOVED = 2,
 }
 
-/**
- * Manages the relationship of this client with the cluster.
- */
-export class ClusterService {
+class MemberListSnapshot {
+    version: number;
+    members: Map<string, Member>;
 
-    /**
-     * The unique identifier of the owner server node. This node is responsible for resource cleanup
-     */
-    public ownerUuid: string = null;
+    constructor(version: number, members: Map<string, Member>) {
+        this.version = version;
+        this.members = members;
+    }
+}
 
-    /**
-     * The unique identifier of this client instance. Assigned by owner node on authentication
-     */
-    public uuid: string = null;
+const EMPTY_SNAPSHOT = new MemberListSnapshot(-1, new Map<string, Member>());
+const INITIAL_MEMBERS_TIMEOUT_SECONDS = 120;
```

Comment:
nit: convert this constant to milliseconds to avoid `INITIAL_MEMBERS_TIMEOUT_SECONDS * 1000` calculations.